### ISSUE TITLE
compilation failure with gcc5.x

### ISSUE THREAD

**Issue by yonghong-song** [iovisor/bpftrace#159]
I hit a compilation error with the following error:
```
-bash-4.2$ make
[  1%] [FLEX][flex_lexer] Building scanner with flex 2.5.37
src/lexer.l:49: warning, rule cannot be matched
[  2%] [BISON][bison_parser] Building parser with bison 3.0.4
Scanning dependencies of target parser
[  3%] Building CXX object CMakeFiles/parser.dir/parser.tab.cc.o
c++: error: unrecognized command line option ‘-std=c++14’
make[2]: *** [CMakeFiles/parser.dir/parser.tab.cc.o] Error 1
make[1]: *** [CMakeFiles/parser.dir/all] Error 2
make: *** [all] Error 2
-bash-4.2$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/mnt/vol/engshare_nosnap/third-party2/gcc/5.x/centos7-native/c447969/bin/../libexec/gcc/x86_64-redhat-linux-gnu/5.5.0/lto-wrapper
Target: x86_64-redhat-linux-gnu
Configured with: /home/engshare/third-party2/gcc/5.x/src/gcc-5/configure --build=x86_64-redhat-linux-gnu --host=x86_64-redhat-linux-gnu --target=x86_64-redhat-linux-gnu --prefix=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969 --exec-prefix=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969 --bindir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/bin --sbindir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/sbin --sysconfdir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/etc --datadir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/share --includedir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/include --libdir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/lib --libexecdir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/libexec --localstatedir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/var --sharedstatedir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/com --mandir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/share/man --infodir=/home/engshare/third-party2/gcc/5.x/centos7-native/c447969/share/info --with-gmp=/home/engshare/third-party2/gmp/6.0.0/centos7-native/da39a3e --with-mpfr=/home/engshare/third-party2/mpfr/3.1.2/centos7-native/1f53355 --with-mpc=/home/engshare/third-party2/mpc/1.0.2/centos7-native/6a50ac4 --with-isl=/home/engshare/third-party2/isl/0.18/centos7-native/1f53355 --with-pic --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-languages=c,c++,fortran,lto --disable-libgcj --disable-gnu-unique-object --enable-linker-build-id --with-gnu-as --with-gnu-ld --disable-multilib --enable-libstdcxx-time --enable-gnu-indirect-function --enable-plugin
Thread model: posix
gcc version 5.x 20180625 (Facebook) 5.5.0+ (GCC) 
-bash-4.2$
```
maybe we should be relaxing -std=c++14 to allow old distro to use the tool?

**Comment by yonghong-song** [comment]
Correction for myself. The above -std=c++14 is not due to gcc 5.x. It is due to my setup whether cmake chooses an old version of gcc 4.8.5.

But even with correct 5.5.x gcc/g++ compiler, I still get the following compilation erros:
```
[ 15%] Building CXX object src/ast/CMakeFiles/ast.dir/ast.cpp.o
[ 16%] Building CXX object src/ast/CMakeFiles/ast.dir/codegen_llvm.cpp.o
In file included from /home/yhs/work/bpftrace/src/ast/codegen_llvm.cpp:1:
/home/yhs/work/bpftrace/src/bpforc.h:50:48: error: ‘ModuleHandleT’ in ‘class llvm::orc::IRCompileLayer<llvm::orc::RTDyldObjectLinkingLayer, 
llvm::orc::SimpleCompiler>’ does not name a type
   using ModuleHandle = decltype(CompileLayer)::ModuleHandleT;
                                                ^~~~~~~~~~~~~
/home/yhs/work/bpftrace/src/bpforc.h:65:3: error: ‘ModuleHandle’ does not name a type
   ModuleHandle addModule(std::unique_ptr<Module> M) {
   ^~~~~~~~~~~~
/home/yhs/work/bpftrace/src/bpforc.h: In constructor ‘bpftrace::BpfOrc::BpfOrc(llvm::TargetMachine*)’:
/home/yhs/work/bpftrace/src/bpforc.h:55:52: error: no matching function for call to ‘llvm::orc::RTDyldObjectLinkingLayer::RTDyldObjectLinkin
gLayer(bpftrace::BpfOrc::BpfOrc(llvm::TargetMachine*)::<lambda()>)’
       CompileLayer(ObjectLayer, SimpleCompiler(*TM))
                                                    ^
In file included from /home/yhs/work/bpftrace/src/bpforc.h:10,
                 from /home/yhs/work/bpftrace/src/ast/codegen_llvm.cpp:1:
/mnt/vol/engshare/third-party2/llvm-fb/stable/gcc-5-glibc-2.23/03859b5/include/llvm/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.h:240:3: no
te: candidate: ‘llvm::orc::RTDyldObjectLinkingLayer::RTDyldObjectLinkingLayer(llvm::orc::ExecutionSession&, llvm::orc::RTDyldObjectLinkingLa
yer::ResourcesGetter, llvm::orc::RTDyldObjectLinkingLayer::NotifyLoadedFtor, llvm::orc::RTDyldObjectLinkingLayer::NotifyFinalizedFtor)’
   RTDyldObjectLinkingLayer(
   ^~~~~~~~~~~~~~~~~~~~~~~~
/mnt/vol/engshare/third-party2/llvm-fb/stable/gcc-5-glibc-2.23/03859b5/include/llvm/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.h:240:3: no
te:   candidate expects 4 arguments, 1 provided
...
```
I am using latest llvm trunk (8.0), so I suspect this is due to the latest llvm issue.

It would be good if you can fix this issue since I am often running with latest llvm.

**Comment by brendangregg** [comment]
Yes, this sounds like llvm API changes. #149 shows the kind of work to fix this. What I don't know is how exactly: should we ifdef for the llvm versions, switch to the C API, or would an llvm engineer tell us there's a better way? (know any llvm engineers other than @4ast?)

I guess we could start by ifdef'ing the llvm versions.

**Comment by mmarchini** [comment]
Closing since this is covered by #137 and there's a proposed fix in #210, but feel free to reopen if there's anything left to be done here.
